pkg/addon/render.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	cueyaml "cuelang.org/go/encoding/yaml"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	common2 "github.com/oam-dev/kubevela/apis/core.oam.dev/common"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/apis/types"
+	cuemodel "github.com/oam-dev/kubevela/pkg/cue/model"
+	"github.com/oam-dev/kubevela/pkg/cue/model/value"
+	"github.com/oam-dev/kubevela/pkg/multicluster"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	addonutil "github.com/oam-dev/kubevela/pkg/utils/addon"
+)
+
+type addonCueTemplateRender struct {
+	addon     *InstallPackage
+	inputArgs map[string]interface{}
+}
+
+// This func can be used for addon render, supporting render app template and component.
+// Please notice the result will be stored in object parameter, so object must be a pointer type
+func (a addonCueTemplateRender) toObject(cueTemplate string, object interface{}) error {
+	bt, err := json.Marshal(a.inputArgs)
+	if err != nil {
+		return err
+	}
+	var contextFile = strings.Builder{}
+	var paramFile = cuemodel.ParameterFieldName + ": {}"
+	if string(bt) != "null" {
+		paramFile = fmt.Sprintf("%s: %s", cuemodel.ParameterFieldName, string(bt))
+	}
+	// addon metadata context
+	metadataJSON, err := json.Marshal(a.addon.Meta)
+	if err != nil {
+		return err
+	}
+	contextFile.WriteString(fmt.Sprintf("context: metadata: %s\n", string(metadataJSON)))
```

Comment:
will the metadataJSON be empty?

Reply:
No,  empty meta will get `{"name":"","version":"","description":"","icon":"","invisible":false}`